[bug][tdengine] connector-tdengine didn't work cause no suitable driver found exception and NPE

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### What happened

unfortunately I got problem connect to tdengine, I tried to find out the cause, but the reason is not printed.

part of log:
```
2023-12-12 19:57:29,452 ERROR org.apache.seatunnel.engine.server.dag.physical.PhysicalVertex - Job SeaTunnel_Job (786924607960514561), Pipeline: [(1/1)], task: [pipeline-1 [Source[0]-TDengine-src]-SourceTask (1/1)] end with state FAILED and Exception: org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException: ErrorCode:[COMMON-12], ErrorDescription:[Source reader operation failed, such as (open, close) etc...] - get TDengine connection failed:jdbc:TAOS-RS://localhost:6041/test?user=root&password=taosdata
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.open(TDengineSourceReader.java:113)
        at org.apache.seatunnel.engine.server.task.flow.SourceFlowLifeCycle.open(SourceFlowLifeCycle.java:115)
        at org.apache.seatunnel.engine.server.task.SeaTunnelTask.stateProcess(SeaTunnelTask.java:146)
        at org.apache.seatunnel.engine.server.task.SourceSeaTunnelTask.call(SourceSeaTunnelTask.java:89)
        at org.apache.seatunnel.engine.server.TaskExecutionService$BlockingWorker.run(TaskExecutionService.java:613)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

2023-12-12 19:57:29,452 ERROR org.apache.seatunnel.engine.server.dag.physical.SubPlan - Task TaskGroupLocation{jobId=786924607960514561, pipelineId=1, taskGroupId=50000} Failed in Job SeaTunnel_Job (786924607960514561), Pipeline: [(1/1)], Begin to cancel other tasks in this pipeline.
```

Next I changed the try catch block, add cause before re-throwing SQLException
![image](https://github.com/apache/seatunnel/assets/17078980/f835b1a6-b894-4b33-bd13-0259d38368db)

I re-run the job and this time cause is printed on the screen
```
Exception in thread "main" org.apache.seatunnel.core.starter.exception.CommandExecuteException: SeaTunnel job executed failed
        at org.apache.seatunnel.core.starter.seatunnel.command.ClientExecuteCommand.execute(ClientExecuteCommand.java:188)
        at org.apache.seatunnel.core.starter.SeaTunnel.run(SeaTunnel.java:40)
        at org.apache.seatunnel.core.starter.seatunnel.SeaTunnelClient.main(SeaTunnelClient.java:34)
Caused by: org.apache.seatunnel.engine.common.exception.SeaTunnelEngineException: org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException: ErrorCode:[COMMON-12], ErrorDescription:[Source reader operation failed, such as (open, close) etc...] - get TDengine connection failed:jdbc:TAOS-RS://localhost:6041/test?user=root&password=taosdata
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.open(TDengineSourceReader.java:102)
        at org.apache.seatunnel.engine.server.task.flow.SourceFlowLifeCycle.open(SourceFlowLifeCycle.java:115)
        at org.apache.seatunnel.engine.server.task.SeaTunnelTask.stateProcess(SeaTunnelTask.java:146)
        at org.apache.seatunnel.engine.server.task.SourceSeaTunnelTask.call(SourceSeaTunnelTask.java:89)
        at org.apache.seatunnel.engine.server.TaskExecutionService$BlockingWorker.run(TaskExecutionService.java:613)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.sql.SQLException: No suitable driver found for jdbc:TAOS-RS://localhost:6041/test?user=root&password=taosdata
        at java.sql.DriverManager.getConnection(DriverManager.java:689)
        at java.sql.DriverManager.getConnection(DriverManager.java:208)
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.open(TDengineSourceReader.java:100)
        ... 9 more

        at org.apache.seatunnel.engine.client.job.ClientJobProxy.waitForJobComplete(ClientJobProxy.java:122)
        at org.apache.seatunnel.core.starter.seatunnel.command.ClientExecuteCommand.execute(ClientExecuteCommand.java:181)
        ... 2 more
```

this seems to be a simple problem, so I changed the source code again:
![image](https://github.com/apache/seatunnel/assets/17078980/6b896389-3b3f-4aa4-a5cb-d42458bf9b08)

......
run the job and get another exception:
```
Exception in thread "main" org.apache.seatunnel.core.starter.exception.CommandExecuteException: SeaTunnel job executed failed
        at org.apache.seatunnel.core.starter.seatunnel.command.ClientExecuteCommand.execute(ClientExecuteCommand.java:188)
        at org.apache.seatunnel.core.starter.SeaTunnel.run(SeaTunnel.java:40)
        at org.apache.seatunnel.core.starter.seatunnel.SeaTunnelClient.main(SeaTunnelClient.java:34)
Caused by: org.apache.seatunnel.engine.common.exception.SeaTunnelEngineException: org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException: ErrorCode:[COMMON-12], ErrorDescription:[Source reader operation failed, such as (open, close) etc...] - TDengine split read error
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.lambda$pollNext$0(TDengineSourceReader.java:66)
        at java.lang.Iterable.forEach(Iterable.java:75)
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.pollNext(TDengineSourceReader.java:61)
        at org.apache.seatunnel.engine.server.task.flow.SourceFlowLifeCycle.collect(SourceFlowLifeCycle.java:135)
        at org.apache.seatunnel.engine.server.task.SourceSeaTunnelTask.collect(SourceSeaTunnelTask.java:84)
        at org.apache.seatunnel.engine.server.task.SeaTunnelTask.stateProcess(SeaTunnelTask.java:165)
        at org.apache.seatunnel.engine.server.task.SourceSeaTunnelTask.call(SourceSeaTunnelTask.java:89)
        at org.apache.seatunnel.engine.server.TaskExecutionService$BlockingWorker.run(TaskExecutionService.java:613)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.convertDataType(TDengineSourceReader.java:145)
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.read(TDengineSourceReader.java:136)
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.lambda$pollNext$0(TDengineSourceReader.java:64)
        ... 12 more

        at org.apache.seatunnel.engine.client.job.ClientJobProxy.waitForJobComplete(ClientJobProxy.java:122)
        at org.apache.seatunnel.core.starter.seatunnel.command.ClientExecuteCommand.execute(ClientExecuteCommand.java:181)
        ... 2 more
```
......
after debug and found the cause is row columns existing null values, convertDataType will throw NPE if parameter is null

......

coding -> run the job -> and this time I succeed!

### SeaTunnel Version

2.3.2

### SeaTunnel Config

```conf
env {
  execution.parallelism = 1
  job.mode = "BATCH"
}

source {
        TDengine {
          url : "jdbc:TAOS-RS://localhost:6041/"
          username : "root"
          password : "taosdata"
          database : "test"
          stable : "xxxxx"
          lower_bound : "2023-12-12 14:38:05.000"
          upper_bound : "2023-12-12 14:38:16.800"
          result_table_name = "src"
        }
}

sink {
  Console {}
}
```


### Running Command

```shell
bin/seatunnel.sh --config a.conf -e local
```


### Error Exception

```log
2023-12-12 19:57:29,452 ERROR org.apache.seatunnel.engine.server.dag.physical.PhysicalVertex - Job SeaTunnel_Job (786924607960514561), Pipeline: [(1/1)], task: [pipeline-1 [Source[0]-TDengine-src]-SourceTask (1/1)] end with state FAILED and Exception: org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException: ErrorCode:[COMMON-12], ErrorDescription:[Source reader operation failed, such as (open, close) etc...] - get TDengine connection failed:jdbc:TAOS-RS://localhost:6041/test?user=root&password=taosdata
        at org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceReader.open(TDengineSourceReader.java:113)
        at org.apache.seatunnel.engine.server.task.flow.SourceFlowLifeCycle.open(SourceFlowLifeCycle.java:115)
        at org.apache.seatunnel.engine.server.task.SeaTunnelTask.stateProcess(SeaTunnelTask.java:146)
        at org.apache.seatunnel.engine.server.task.SourceSeaTunnelTask.call(SourceSeaTunnelTask.java:89)
        at org.apache.seatunnel.engine.server.TaskExecutionService$BlockingWorker.run(TaskExecutionService.java:613)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```


### Zeta or Flink or Spark Version

_No response_

### Java or Scala Version

_No response_

### Screenshots

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/config/TDengineSourceConfig.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/config/TDengineSourceConfig.java
@@ -30,7 +30,6 @@ import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengine
 import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig.ConfigNames.STABLE;
 import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig.ConfigNames.TIMEZONE;
 import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig.ConfigNames.UPPER_BOUND;
-import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig.ConfigNames.URL;
 import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig.ConfigNames.USERNAME;
 
 @Data
@@ -54,7 +53,10 @@ public class TDengineSourceConfig implements Serializable {
 
     public static TDengineSourceConfig buildSourceConfig(Config pluginConfig) {
         TDengineSourceConfig tdengineSourceConfig = new TDengineSourceConfig();
-        tdengineSourceConfig.setUrl(pluginConfig.hasPath(URL) ? pluginConfig.getString(URL) : null);
+        tdengineSourceConfig.setUrl(
+                pluginConfig.hasPath(ConfigNames.URL)
+                        ? pluginConfig.getString(ConfigNames.URL)
+                        : null);
         tdengineSourceConfig.setDatabase(
                 pluginConfig.hasPath(DATABASE) ? pluginConfig.getString(DATABASE) : null);
         tdengineSourceConfig.setStable(
@@ -69,6 +71,7 @@ public class TDengineSourceConfig implements Serializable {
                 pluginConfig.hasPath(LOWER_BOUND) ? pluginConfig.getString(LOWER_BOUND) : null);
         tdengineSourceConfig.setTimezone(
                 pluginConfig.hasPath(TIMEZONE) ? pluginConfig.getString(TIMEZONE) : "UTC");
+
         return tdengineSourceConfig;
     }
 

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSource.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSource.java
@@ -40,6 +40,7 @@ import org.apache.seatunnel.connectors.seatunnel.tdengine.typemapper.TDengineTyp
 import org.apache.commons.lang3.ArrayUtils;
 
 import com.google.auto.service.AutoService;
+import com.taosdata.jdbc.TSDBDriver;
 import lombok.SneakyThrows;
 
 import java.sql.Connection;
@@ -49,6 +50,7 @@ import java.sql.SQLException;
 import java.sql.Statement;
 import java.util.ArrayList;
 import java.util.List;
+import java.util.Properties;
 
 import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig.ConfigNames.DATABASE;
 import static org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig.ConfigNames.PASSWORD;
@@ -127,42 +129,36 @@ public class TDengineSource
         List<String> fieldNames = new ArrayList<>();
         List<SeaTunnelDataType<?>> fieldTypes = new ArrayList<>();
 
-        String jdbcUrl =
-                String.join(
-                        "",
-                        config.getUrl(),
-                        config.getDatabase(),
-                        "?user=",
-                        config.getUsername(),
-                        "&password=",
-                        config.getPassword());
+        String jdbcUrl = String.join("", config.getUrl(), config.getDatabase());
+
         // check td driver whether exist and if not, try to register
         checkDriverExist(jdbcUrl);
-        try (Connection conn = DriverManager.getConnection(jdbcUrl)) {
-            try (Statement statement = conn.createStatement()) {
+
+        Properties properties = new Properties();
+        properties.put(TSDBDriver.PROPERTY_KEY_USER, config.getUsername());
+        properties.put(TSDBDriver.PROPERTY_KEY_PASSWORD, config.getPassword());
+        String metaSQL =
+                String.format(
+                        "select table_name from information_schema.ins_tables where db_name = '%s' and stable_name='%s'",
+                        config.getDatabase(), config.getStable());
+        try (Connection conn = DriverManager.getConnection(jdbcUrl, properties);
+                Statement statement = conn.createStatement();
                 ResultSet metaResultSet =
                         statement.executeQuery(
-                                "desc " + config.getDatabase() + "." + config.getStable());
-                while (metaResultSet.next()) {
-                    if (timestampFieldName == null) {
-                        timestampFieldName = metaResultSet.getString(1);
-                    }
-                    fieldNames.add(metaResultSet.getString(1));
-                    fieldTypes.add(TDengineTypeMapper.mapping(metaResultSet.getString(2)));
+                                String.format(
+                                        "desc %s.%s", config.getDatabase(), config.getStable()));
+                ResultSet subTableNameResultSet = statement.executeQuery(metaSQL)) {
+            while (metaResultSet.next()) {
+                if (timestampFieldName == null) {
+                    timestampFieldName = metaResultSet.getString(1);
                 }
+                fieldNames.add(metaResultSet.getString(1));
+                fieldTypes.add(TDengineTypeMapper.mapping(metaResultSet.getString(2)));
             }
-            try (Statement statement = conn.createStatement()) {
-                String metaSQL =
-                        "select table_name from information_schema.ins_tables where db_name = '"
-                                + config.getDatabase()
-                                + "' and stable_name='"
-                                + config.getStable()
-                                + "';";
-                ResultSet subTableNameResultSet = statement.executeQuery(metaSQL);
-                while (subTableNameResultSet.next()) {
-                    String subTableName = subTableNameResultSet.getString(1);
-                    subTableNames.add(subTableName);
-                }
+
+            while (subTableNameResultSet.next()) {
+                String subTableName = subTableNameResultSet.getString(1);
+                subTableNames.add(subTableName);
             }
         }
 

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
@@ -81,7 +81,7 @@ public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengine
                 }
             } else if (noMoreSplit && sourceSplits.isEmpty()) {
                 // signal to the source that we have reached the end of the data.
-                log.info("Closed the bounded jdbc source");
+                log.info("Closed the bounded TDengine source");
                 context.signalNoMoreElement();
             } else {
                 Thread.sleep(1000L);

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
@@ -17,7 +17,6 @@
 
 package org.apache.seatunnel.connectors.seatunnel.tdengine.source;
 
-import com.taosdata.jdbc.TSDBDriver;
 import org.apache.seatunnel.api.source.Collector;
 import org.apache.seatunnel.api.source.SourceReader;
 import org.apache.seatunnel.api.table.type.SeaTunnelRow;
@@ -25,8 +24,7 @@ import org.apache.seatunnel.common.exception.CommonErrorCodeDeprecated;
 import org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig;
 import org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException;
 
-import org.apache.commons.lang3.StringUtils;
-
+import com.taosdata.jdbc.TSDBDriver;
 import lombok.extern.slf4j.Slf4j;
 
 import java.sql.Connection;

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
@@ -123,8 +123,8 @@ public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengine
     }
 
     private void read(TDengineSourceSplit split, Collector<SeaTunnelRow> output) throws Exception {
-        try (Statement statement = conn.createStatement()) {
-            final ResultSet resultSet = statement.executeQuery(split.getQuery());
+        try (Statement statement = conn.createStatement();
+                ResultSet resultSet = statement.executeQuery(split.getQuery())) {
             ResultSetMetaData meta = resultSet.getMetaData();
 
             while (resultSet.next()) {

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
@@ -114,7 +114,8 @@ public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengine
         } catch (SQLException e) {
             throw new TDengineConnectorException(
                     CommonErrorCodeDeprecated.READER_OPERATION_FAILED,
-                    "get TDengine connection failed:" + jdbcUrl, e);
+                    "get TDengine connection failed:" + jdbcUrl,
+                    e);
         }
     }
 

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
@@ -17,6 +17,7 @@
 
 package org.apache.seatunnel.connectors.seatunnel.tdengine.source;
 
+import com.taosdata.jdbc.TSDBDriver;
 import org.apache.seatunnel.api.source.Collector;
 import org.apache.seatunnel.api.source.SourceReader;
 import org.apache.seatunnel.api.table.type.SeaTunnelRow;
@@ -26,7 +27,6 @@ import org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConn
 
 import org.apache.commons.lang3.StringUtils;
 
-import com.taosdata.jdbc.TSDBDriver;
 import lombok.extern.slf4j.Slf4j;
 
 import java.sql.Connection;
@@ -93,24 +93,15 @@ public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengine
 
     @Override
     public void open() {
-        String jdbcUrl =
-                StringUtils.join(
-                        config.getUrl(),
-                        config.getDatabase(),
-                        "?user=",
-                        config.getUsername(),
-                        "&password=",
-                        config.getPassword());
-        Properties connProps = new Properties();
-        // todo: when TSDBDriver.PROPERTY_KEY_BATCH_LOAD set to "true",
-        // there is a exception : Caused by: java.sql.SQLException: can't create connection with
-        // server
-        // under docker network env
-        // @bobo (tdengine)
-        connProps.setProperty(TSDBDriver.PROPERTY_KEY_BATCH_LOAD, "true");
+        String jdbcUrl = config.getUrl();
+
+        Properties properties = new Properties();
+        properties.put(TSDBDriver.PROPERTY_KEY_USER, config.getUsername());
+        properties.put(TSDBDriver.PROPERTY_KEY_PASSWORD, config.getPassword());
+
         try {
             checkDriverExist(jdbcUrl);
-            conn = DriverManager.getConnection(jdbcUrl, connProps);
+            conn = DriverManager.getConnection(jdbcUrl, properties);
         } catch (SQLException e) {
             throw new TDengineConnectorException(
                     CommonErrorCodeDeprecated.READER_OPERATION_FAILED,

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReader.java
@@ -17,7 +17,6 @@
 
 package org.apache.seatunnel.connectors.seatunnel.tdengine.source;
 
-import org.apache.seatunnel.api.source.Boundedness;
 import org.apache.seatunnel.api.source.Collector;
 import org.apache.seatunnel.api.source.SourceReader;
 import org.apache.seatunnel.api.table.type.SeaTunnelRow;
@@ -27,7 +26,6 @@ import org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConn
 
 import org.apache.commons.lang3.StringUtils;
 
-import com.google.common.collect.Sets;
 import com.taosdata.jdbc.TSDBDriver;
 import lombok.extern.slf4j.Slf4j;
 
@@ -39,56 +37,57 @@ import java.sql.SQLException;
 import java.sql.Statement;
 import java.sql.Timestamp;
 import java.util.ArrayList;
+import java.util.Deque;
 import java.util.List;
 import java.util.Objects;
 import java.util.Properties;
-import java.util.Set;
+import java.util.concurrent.ConcurrentLinkedDeque;
 
 import static org.apache.seatunnel.connectors.seatunnel.tdengine.utils.TDengineUtil.checkDriverExist;
 
 @Slf4j
 public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengineSourceSplit> {
-
-    private static final long THREAD_WAIT_TIME = 500L;
-
     private final TDengineSourceConfig config;
 
-    private final Set<TDengineSourceSplit> sourceSplits;
+    private final Deque<TDengineSourceSplit> sourceSplits;
 
     private final Context context;
 
     private Connection conn;
 
+    private volatile boolean noMoreSplit;
+
     public TDengineSourceReader(TDengineSourceConfig config, SourceReader.Context readerContext) {
         this.config = config;
-        this.sourceSplits = Sets.newHashSet();
+        this.sourceSplits = new ConcurrentLinkedDeque<>();
         this.context = readerContext;
     }
 
     @Override
     public void pollNext(Collector<SeaTunnelRow> collector) throws InterruptedException {
-        if (sourceSplits.isEmpty()) {
-            Thread.sleep(THREAD_WAIT_TIME);
-            return;
-        }
         synchronized (collector.getCheckpointLock()) {
-            sourceSplits.forEach(
-                    split -> {
-                        try {
-                            read(split, collector);
-                        } catch (Exception e) {
-                            throw new TDengineConnectorException(
-                                    CommonErrorCodeDeprecated.READER_OPERATION_FAILED,
-                                    "TDengine split read error",
-                                    e);
-                        }
-                    });
-        }
-
-        if (Boundedness.BOUNDED.equals(context.getBoundedness())) {
-            // signal to the source that we have reached the end of the data.
-            log.info("Closed the bounded TDengine source");
-            context.signalNoMoreElement();
+            log.info("polling new split from queue!");
+            TDengineSourceSplit split = sourceSplits.poll();
+            if (Objects.nonNull(split)) {
+                log.info(
+                        "starting run new split {}, query sql: {}!",
+                        split.splitId(),
+                        split.getQuery());
+                try {
+                    read(split, collector);
+                } catch (Exception e) {
+                    throw new TDengineConnectorException(
+                            CommonErrorCodeDeprecated.READER_OPERATION_FAILED,
+                            "TDengine split read error",
+                            e);
+                }
+            } else if (noMoreSplit && sourceSplits.isEmpty()) {
+                // signal to the source that we have reached the end of the data.
+                log.info("Closed the bounded jdbc source");
+                context.signalNoMoreElement();
+            } else {
+                Thread.sleep(1000L);
+            }
         }
     }
 
@@ -108,15 +107,14 @@ public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengine
         // server
         // under docker network env
         // @bobo (tdengine)
-        connProps.setProperty(TSDBDriver.PROPERTY_KEY_BATCH_LOAD, "false");
+        connProps.setProperty(TSDBDriver.PROPERTY_KEY_BATCH_LOAD, "true");
         try {
-            // check td driver whether exist and if not, try to register
             checkDriverExist(jdbcUrl);
             conn = DriverManager.getConnection(jdbcUrl, connProps);
         } catch (SQLException e) {
             throw new TDengineConnectorException(
                     CommonErrorCodeDeprecated.READER_OPERATION_FAILED,
-                    "get TDengine connection failed:" + jdbcUrl);
+                    "get TDengine connection failed:" + jdbcUrl, e);
         }
     }
 
@@ -151,6 +149,8 @@ public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengine
     }
 
     private Object convertDataType(Object object) {
+        if (Objects.isNull(object)) return null;
+
         if (Timestamp.class.equals(object.getClass())) {
             return ((Timestamp) object).toLocalDateTime();
         } else if (byte[].class.equals(object.getClass())) {
@@ -171,7 +171,8 @@ public class TDengineSourceReader implements SourceReader<SeaTunnelRow, TDengine
 
     @Override
     public void handleNoMoreSplits() {
-        // do nothing
+        log.info("no more split accepted!");
+        noMoreSplit = true;
     }
 
     @Override

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceSplitEnumerator.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceSplitEnumerator.java
@@ -142,7 +142,7 @@ public class TDengineSourceSplitEnumerator
 
     @Override
     public void addSplitsBack(List<TDengineSourceSplit> splits, int subtaskId) {
-        log.info("Add back splits {} to JdbcSourceSplitEnumerator.", splits);
+        log.info("Add back splits {} to TDengineSourceSplitEnumerator.", splits);
         if (!splits.isEmpty()) {
             addPendingSplit(splits);
             assignSplit(Collections.singletonList(subtaskId));

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceSplitEnumerator.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceSplitEnumerator.java
@@ -17,28 +17,34 @@
 
 package org.apache.seatunnel.connectors.seatunnel.tdengine.source;
 
-import org.apache.seatunnel.api.source.SourceEvent;
 import org.apache.seatunnel.api.source.SourceSplitEnumerator;
+import org.apache.seatunnel.common.exception.CommonErrorCodeDeprecated;
 import org.apache.seatunnel.connectors.seatunnel.tdengine.config.TDengineSourceConfig;
+import org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException;
 import org.apache.seatunnel.connectors.seatunnel.tdengine.state.TDengineSourceState;
 
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collection;
 import java.util.Collections;
-import java.util.HashSet;
 import java.util.List;
-import java.util.Objects;
+import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.stream.Collectors;
 
+@Slf4j
 public class TDengineSourceSplitEnumerator
         implements SourceSplitEnumerator<TDengineSourceSplit, TDengineSourceState> {
 
     private final SourceSplitEnumerator.Context<TDengineSourceSplit> context;
     private final TDengineSourceConfig config;
     private final StableMetadata stableMetadata;
-    private Set<TDengineSourceSplit> pendingSplit = new HashSet<>();
-    private Set<TDengineSourceSplit> assignedSplit = new HashSet<>();
+    private volatile boolean shouldEnumerate;
+    private final Object stateLock = new Object();
+    private final Map<Integer, List<TDengineSourceSplit>> pendingSplits = new ConcurrentHashMap<>();
 
     public TDengineSourceSplitEnumerator(
             StableMetadata stableMetadata,
@@ -55,8 +61,10 @@ public class TDengineSourceSplitEnumerator
         this.config = config;
         this.context = context;
         this.stableMetadata = stableMetadata;
+        this.shouldEnumerate = sourceState == null;
         if (sourceState != null) {
-            this.assignedSplit = sourceState.getAssignedSplit();
+            this.shouldEnumerate = sourceState.isShouldEnumerate();
+            this.pendingSplits.putAll(sourceState.getPendingSplits());
         }
     }
 
@@ -69,16 +77,33 @@ public class TDengineSourceSplitEnumerator
 
     @Override
     public void run() {
-        pendingSplit = getAllSplits();
-        assignSplit(context.registeredReaders());
+        Set<Integer> readers = context.registeredReaders();
+        if (shouldEnumerate) {
+            List<TDengineSourceSplit> newSplits = discoverySplits();
+
+            synchronized (stateLock) {
+                addPendingSplit(newSplits);
+                shouldEnumerate = false;
+            }
+
+            assignSplit(readers);
+        }
+
+        log.info("No more splits to assign." + " Sending NoMoreSplitsEvent to reader {}.", readers);
+        readers.forEach(context::signalNoMoreSplits);
     }
 
-    /*
-     * each split has one sub table
-     */
-    private Set<TDengineSourceSplit> getAllSplits() {
+    private void addPendingSplit(List<TDengineSourceSplit> newSplits) {
+        int readerCount = context.currentParallelism();
+        for (TDengineSourceSplit split : newSplits) {
+            int ownerReader = getSplitOwner(split.splitId(), readerCount);
+            pendingSplits.computeIfAbsent(ownerReader, r -> new ArrayList<>()).add(split);
+        }
+    }
+
+    private List<TDengineSourceSplit> discoverySplits() {
         final String timestampFieldName = stableMetadata.getTimestampFieldName();
-        final Set<TDengineSourceSplit> splits = new HashSet<>();
+        final List<TDengineSourceSplit> splits = new ArrayList<>();
         for (String subTableName : stableMetadata.getSubTableNames()) {
             TDengineSourceSplit splitBySubTable =
                     createSplitBySubTable(subTableName, timestampFieldName);
@@ -92,6 +117,7 @@ public class TDengineSourceSplitEnumerator
         String selectFields =
                 Arrays.stream(stableMetadata.getRowType().getFieldNames())
                         .skip(1)
+                        .map(name -> String.format("`%s`", name))
                         .collect(Collectors.joining(","));
         String subTableSQL =
                 "select " + selectFields + " from " + config.getDatabase() + "." + subTableName;
@@ -116,69 +142,64 @@ public class TDengineSourceSplitEnumerator
 
     @Override
     public void addSplitsBack(List<TDengineSourceSplit> splits, int subtaskId) {
+        log.info("Add back splits {} to JdbcSourceSplitEnumerator.", splits);
         if (!splits.isEmpty()) {
-            pendingSplit.addAll(splits);
+            addPendingSplit(splits);
             assignSplit(Collections.singletonList(subtaskId));
         }
     }
 
     @Override
     public int currentUnassignedSplitSize() {
-        return pendingSplit.size();
+        return pendingSplits.size();
     }
 
     @Override
     public void registerReader(int subtaskId) {
-        if (!pendingSplit.isEmpty()) {
+        log.info("Register reader {} to TDengineSourceSplitEnumerator.", subtaskId);
+        if (!pendingSplits.isEmpty()) {
             assignSplit(Collections.singletonList(subtaskId));
         }
     }
 
-    private void assignSplit(Collection<Integer> taskIDList) {
-        assignedSplit =
-                pendingSplit.stream()
-                        .map(
-                                split -> {
-                                    int splitOwner =
-                                            getSplitOwner(
-                                                    split.splitId(), context.currentParallelism());
-                                    if (taskIDList.contains(splitOwner)) {
-                                        context.assignSplit(splitOwner, split);
-                                        return split;
-                                    } else {
-                                        return null;
-                                    }
-                                })
-                        .filter(Objects::nonNull)
-                        .collect(Collectors.toSet());
-        pendingSplit.clear();
+    private void assignSplit(Collection<Integer> readers) {
+        log.info("Assign pendingSplits to readers {}", readers);
+
+        for (int reader : readers) {
+            List<TDengineSourceSplit> assignmentForReader = pendingSplits.remove(reader);
+            if (assignmentForReader != null && !assignmentForReader.isEmpty()) {
+                log.info("Assign splits {} to reader {}", assignmentForReader, reader);
+                try {
+                    context.assignSplit(reader, assignmentForReader);
+                } catch (Exception e) {
+                    log.error(
+                            "Failed to assign splits {} to reader {}",
+                            assignmentForReader,
+                            reader,
+                            e);
+                    pendingSplits.put(reader, assignmentForReader);
+                }
+            }
+        }
     }
 
     @Override
     public TDengineSourceState snapshotState(long checkpointId) {
-        return new TDengineSourceState(assignedSplit);
+        synchronized (stateLock) {
+            return new TDengineSourceState(shouldEnumerate, pendingSplits);
+        }
     }
 
     @Override
-    public void handleSourceEvent(int subtaskId, SourceEvent sourceEvent) {
-        SourceSplitEnumerator.super.handleSourceEvent(subtaskId, sourceEvent);
-    }
-
-    @Override
-    public void notifyCheckpointComplete(long checkpointId) {
-        // nothing to do
-    }
-
-    @Override
-    public void notifyCheckpointAborted(long checkpointId) throws Exception {
-        SourceSplitEnumerator.super.notifyCheckpointAborted(checkpointId);
-    }
+    public void notifyCheckpointComplete(long checkpointId) {}
 
     @Override
     public void close() {}
 
     @Override
     public void handleSplitRequest(int subtaskId) {
-        // nothing to do
+        throw new TDengineConnectorException(
+                CommonErrorCodeDeprecated.UNSUPPORTED_OPERATION,
+                String.format("Unsupported handleSplitRequest: %d", subtaskId));
     }
 }

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceSplitEnumerator.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceSplitEnumerator.java
@@ -120,7 +120,8 @@ public class TDengineSourceSplitEnumerator
                         .map(name -> String.format("`%s`", name))
                         .collect(Collectors.joining(","));
         String subTableSQL =
-                "select " + selectFields + " from " + config.getDatabase() + "." + subTableName;
+                String.format(
+                        "select %s from %s.`%s`", selectFields, config.getDatabase(), subTableName);
         String start = config.getLowerBound();
         String end = config.getUpperBound();
         if (start != null || end != null) {

--- a/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/state/TDengineSourceState.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/main/java/org/apache/seatunnel/connectors/seatunnel/tdengine/state/TDengineSourceState.java
@@ -19,18 +19,16 @@ package org.apache.seatunnel.connectors.seatunnel.tdengine.state;
 
 import org.apache.seatunnel.connectors.seatunnel.tdengine.source.TDengineSourceSplit;
 
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+
 import java.io.Serializable;
-import java.util.Set;
+import java.util.List;
+import java.util.Map;
 
+@AllArgsConstructor
+@Getter
 public class TDengineSourceState implements Serializable {
-
-    private final Set<TDengineSourceSplit> assignedSplit;
-
-    public TDengineSourceState(Set<TDengineSourceSplit> assignedSplit) {
-        this.assignedSplit = assignedSplit;
-    }
-
-    public Set<TDengineSourceSplit> getAssignedSplit() {
-        return assignedSplit;
-    }
+    private boolean shouldEnumerate;
+    private final Map<Integer, List<TDengineSourceSplit>> pendingSplits;
 }

--- a/seatunnel-connectors-v2/connector-tdengine/src/test/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReaderTest.java
+++ b/seatunnel-connectors-v2/connector-tdengine/src/test/java/org/apache/seatunnel/connectors/seatunnel/tdengine/source/TDengineSourceReaderTest.java
@@ -1,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.seatunnel.connectors.seatunnel.tdengine.source;
+
+import org.apache.seatunnel.api.source.Collector;
+import org.apache.seatunnel.api.table.type.SeaTunnelRow;
+import org.apache.seatunnel.connectors.seatunnel.tdengine.exception.TDengineConnectorException;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+
+class TDengineSourceReaderTest {
+    Logger logger;
+    TDengineSourceReader tDengineSourceReader;
+
+    @BeforeEach
+    void setup() {
+        tDengineSourceReader = new TDengineSourceReader(null, null);
+
+        List<TDengineSourceSplit> sourceSplits = new ArrayList<>();
+        int splitCnt = 100;
+        for (int i = 0; i < splitCnt; i++) {
+            sourceSplits.add(new TDengineSourceSplit(Integer.toString(i), "select sever_status()"));
+        }
+
+        tDengineSourceReader.addSplits(sourceSplits);
+
+        logger = Logger.getLogger("TDengineSourceReaderTest");
+    }
+
+    @Test
+    void testPoll() throws InterruptedException {
+        TestCollector testCollector = new TestCollector();
+
+        int totalSplitCnt = 150;
+        ThreadPoolExecutor pool =
+                new ThreadPoolExecutor(8, 8, 60, TimeUnit.SECONDS, new LinkedBlockingQueue<>());
+        pool.execute(
+                () -> {
+                    for (int i = 0; i < totalSplitCnt; i++) {
+                        try {
+                            tDengineSourceReader.pollNext(testCollector);
+                            Thread.sleep(new Random().nextInt(5));
+                        } catch (TDengineConnectorException e) {
+                            logger.info("skip create connection!");
+                        } catch (InterruptedException e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                });
+
+        int newSplitCnt = 50;
+        int threadCnt = 3;
+        for (int i = 0; i < threadCnt; i++) {
+            pool.execute(
+                    () -> {
+                        for (int idx = 0; idx < newSplitCnt; idx++) {
+                            logger.info(
+                                    String.format(
+                                            "%s receive new split",
+                                            Thread.currentThread().getName()));
+                            tDengineSourceReader.addSplits(
+                                    Collections.singletonList(
+                                            new TDengineSourceSplit(
+                                                    String.format(
+                                                            "new_%s",
+                                                            Thread.currentThread().getName() + idx),
+                                                    "select server_status()")));
+                            try {
+                                Thread.sleep(new Random().nextInt(5));
+                            } catch (InterruptedException e) {
+                                throw new RuntimeException(e);
+                            }
+                        }
+                    });
+        }
+
+        pool.awaitTermination(3, TimeUnit.SECONDS);
+    }
+
+    private static class TestCollector implements Collector<SeaTunnelRow> {
+
+        private final List<SeaTunnelRow> rows = new ArrayList<>();
+
+        public List<SeaTunnelRow> getRows() {
+            return rows;
+        }
+
+        @Override
+        public void collect(SeaTunnelRow record) {
+            rows.add(record);
+        }
+
+        @Override
+        public Object getCheckpointLock() {
+            return new Object();
+        }
+    }
+}

